src/main/java/com/skirlez/fabricatedexchange/item/tools/RedMatterAxe.java

```diff
@@ -0,0 +1,189 @@
+import net.minecraft.world.World;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class RedMatterAxe extends AxeItem implements ChargeableItem, OutliningItem, ItemWithModes {
+
+    public RedMatterAxe(ToolMaterial material, float attackDamage, float attackSpeed, Settings settings) {
+        super(material, attackDamage, attackSpeed, settings);
+    }
+
+    @Override
+    public int getModeAmount() {
+        return 2;
+    }
+
+    @Override
+    public boolean modeSwitchCondition(ItemStack stack) {
+        return ChargeableItem.getCharge(stack) != 0;
+    }
+
+    @Override
+    public boolean isItemBarVisible(ItemStack stack) {
+        return true;
+    }
+
+    @Override
+    public int getItemBarColor(ItemStack stack) {
+        return ChargeableItem.COLOR;
+    }
+
+    @Override
+    public int getItemBarStep(ItemStack stack) {
+        return ChargeableItem.getItemBarStep(stack, getMaxCharge());
+    }
+
+    @Override
+    public int getMaxCharge() {
+        return 3;
+    }
+
+    @Override
+    public ActionResult useOnBlock(ItemUsageContext context) {
+        int charge = ChargeableItem.getCharge(context.getStack());
+        if (charge == 0)
+            return super.useOnBlock(context);
+
+
+        boolean anySuccess = false;
+        List<BlockPos> positions = getPositionsToOutline(context.getPlayer(), context.getStack(), context.getBlockPos());
+        for (BlockPos newPos : positions) {
+            FakeItemUsageContext fakeContext = 
+                new FakeItemUsageContext(context.getPlayer(), context.getHand(), newPos, Direction.WEST);
+            anySuccess = (super.useOnBlock(fakeContext).isAccepted()) || anySuccess;
+        }
+
+        return ActionResult.success(anySuccess);
+    }
+
+    @Override
+    public float getMiningSpeedMultiplier(ItemStack stack, BlockState state) {
+        float value = super.getMiningSpeedMultiplier(stack, state);
+        if (isSuitableFor(state)) {
+            value += ChargeableItem.getCharge(stack) * 15; // Increase the multiplier compared to dark matter
+        }
+        return value;
+    }
+
+    @Override
+    public boolean postMine(ItemStack stack, World world, BlockState state, BlockPos pos, LivingEntity miner) {
+        // Check if we're on the server side and the block is suitable.
+        if (!world.isClient && miner instanceof PlayerEntity && isSuitableFor(state)) {
+            PlayerEntity player = (PlayerEntity) miner;
+
+            // Find the first suitable block around the initial block.
+            BlockPos firstSuitablePos = findFirstSuitableBlockAround(world, pos);
+            if (firstSuitablePos != null) {
+                // Get the positions of all connected logs from the first found suitable block.
+                List<BlockPos> positions = getPositionsToOutline(player, stack, firstSuitablePos);
+                FabricatedExchange.LOGGER.info("Connected Logs: {}", positions);
```

Comment:
This actually causes issues if you mine anywhere from a log that isn't the start or the end. For example, mining the center log in a tower of 3 blocks will cause nothing else to be mined even though they're outlined, because the firstSuitablePos cannot find any logs...